### ISSUE TITLE
행안부 데이터 기반 폐업률 예측 모델 개발

### ISSUE THREAD

**Issue by syleeie2310** [da-analysis/asac_7_dataanalysis#54]
행안부 데이터 기반 폐업률 예측 모델 개발

**Comment by khye00** [comment]
https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/2390006029154672?o=639069795658224#command/5573037785125618

**Comment by syleeie2310** [comment]
좌표 데이터가 있는 경우에, 법정동 중심 좌표 기준으로 시도/시군구/법정동을 채워넣기
1. geoPandas(https://velog.io/@waterglasses_k/python%EB%B2%95%EC%A0%95%EB%8F%99-%EC%A4%91%EC%8B%AC%EC%A0%90-%EC%A2%8C%ED%91%9C-%EA%B5%AC%ED%95%98%EA%B8%B0)
2. 소진공 데이터랑 조인해서 시도/시군구/법정동 채워넣기
-> 좌표만 join해서 시도/시군구/법정동 unique한 값 채워넣기 (좌표를 join할 때 어느정도 레벨로 할지 결정 필요, 소수점 몇자리)

**Comment by syleeie2310** [comment]
좌표가 없는 경우
1. 수집했을 때 해당 주소가 있는 경우, 없는 경우가 있다. api마다 수준이 다르다.
----> 통계청, 행안부 . 통계청 api로 테스트 완료.

**Comment by syleeie2310** [comment]
https://business.juso.go.kr/addrlink/openApi/apiReqst.do

**Comment by syleeie2310** [comment]
업태구분명 업체 숫자 기준으로 5개 이상 혹은 10개 이상 기타로 묶을지? 판단 필요

**Comment by syleeie2310** [comment]
사업장명 -> 임베딩 방법 고민 필요

https://github.com/Seondong/LocEmb

**Comment by khye00** [comment]
[현재 진행 상황]

1. 시도/시군구/읍면동 값 채우기
2. 운영기간 범주화 작업
3. 읍면동 라벨링 작업
4. 업종 세부분류 작업

**Comment by syleeie2310** [comment]
- silver.mois.mois_geo_v2 (행안부 좌표 -> 법정동 등 정보를 채운 테이블)
- silver.mois.mois_address_info_v2 (좌표가 없는 경우 행안부 api 통해서 -> 법정동 등 정보를 채운 테이블)
- silver.mois.mois_geo_v3 (위 2개 결합한 테이블)

**Comment by syleeie2310** [comment]
- label(폐업 유무 - 폐업일자 기준으로 최근 5년 존재하면 1, 아니면 0)
- 영업상태명
- 영업상태구분코드
- 상세영업상태명
- 상세영업상태코드
- 폐업일자
- 위생업태명
- 사업장명
- 인허가 일자
- 시도, 시군구, 법정동
- 운영 기간
- 업종 (개방서비스명/업태구분명)
- 최종수정시점 -> 날짜로 변경
- 데이터갱신일자 -> 날짜로 변경
- 데이터갱신구분
- 소재지면적

1) 델타 테이블 만들어 주시고, 분류기 -> 훈련/테스트 데이터를 인허가 일자 기준으로 데이터를 나눠야 될것 같음. 어떻게 나눌지?
2) 훈련/테스트 데이터를 나눈 다음에 훈련 데이터 -> 샘플링 -> 모델 학습. **훈련 데이터 90% / 테스트 데이터 10%**
3) 샘플링 1백만개 -> 모델 1차. (개방서비스명/시도/시군구/운영기간) ---> scikit-learn (random forest, xgboost ---> catboost)
(https://catboost.ai/)
4) 1천만개까지 늘린다고 하면 법정동 + 업태구분명 + 다른 변수를 사용할지 고민해야 되는데... ---> 서버리스에서 실험이 안될거라 ---> spark.mllib 구현해야 되서.

**Comment by khye00** [comment]
@syleeie2310 
멘토님 안녕하세요. 델타 테이블 생성 완료했습니다!!!
 - 테이블 명 : silver.mois.mois_deltatable

**Comment by syleeie2310** [comment]
@khye00 고생하셨어요 데이터 갯수가 5백9십만개 정도 있군요.

- 4,125,083
- 1,826,754

보니깐 인허가 일자로 나누면 안될거 같긴 하네요. 인허가 날짜 기준으로 아래처럼 되어 있는거 같아서요. 

데이터 어떻게 나눠서 훈련 데이터 만들지 해당 데이터로 고민해보시겠어요? 샘플링 해서 쓰는 방향으로 할지 

아님 계층적 샘플링 방식으로 개방서비스명 기준으로 나눠도 좋을 것 같습니다.


<img width="861" alt="Image" src="https://github.com/user-attachments/assets/14b26e6c-5c00-4c6f-9255-df0e52ac2eb9" />

**Comment by khye00** [comment]
네 멘토님 고민해보겠습니다.

**Comment by khye00** [comment]
멘토님!!!
폐업일자를 기준으로 확인했을 때 폐업일자가 전부 없는 업종이 15개 있는데,
영업상태명으로 확인해 보았을 때도 '폐업'이 없는 업종이 있습니다.

그래서 폐업일자가 전부 없는 업종 15개는 제외하고 180개 업종으로 진행해야 할지,
폐업수를 기준으로 상위 30~50개 업종만 모델링 진행하는 것이 좋을지 궁금합니다.

[행안부 폐업관련 데이터브릭스 노트북](https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/4113467780411066?o=639069795658224#command/5718355666898378)

**Comment by syleeie2310** [comment]
@khye00 폐업일자가 전부 없는 업종 15개는 제외해도 될거 같고요.

폐업수를 기준으로 상위 몇개만 선택할거면 데이터 갯수 보고 판단하심 좋을 것 같은데. 상위 50개로 하면 너무 줄이는거 같아서

총합 100개 이상 중에서 폐업 숫자가 있는 경우를 사용하면 어떨까 싶습니다.

**Comment by seoyoung1027** [comment]
train test 분할 방법 1(초반에 생각했던 방향), 분할 방법 2(멘토님 조언에 따라 진행)를 참고하여 train test 분할 후 클래스 비율을 확인해보았습니다. 멘토님 조언을 참고하여 업종 갯수를 더 많이 가져가는 2번으로 진행하면 될지 궁금합니다. 

[train/test 비율](https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/4113467780411078?o=639069795658224#command/4113467780411086)

### 방법 1
1. 업종 필터링
- (폐업률 >= 20% & 폐업수 >= 1000)를 적용하여 업종 20개 선택 
- 이유 : 폐업률은 큰데 폐업수가 적은 업종이 존재 (각 업종별로 다르고, 업종 전체 데이터 자체가 작은 경우 존재) 
2. 폐업일자 기준 최근 5년 존재하면 (2020~2025년) 1 (폐업), 아니면 0 (운영)으로 라벨링 
3. 업종 필터링, 계층적 샘플링 진행(비율 유지를 위해) 
4. 결과 (80:20) 
 - 0 : 1192964 
 - 1 : 538042 
- test 
 - 0 : 298231 
 - 1 : 134503 


### 방법 2
1. 폐업일자 기준 폐업이 없고, 총합이 100개 이상인 업종만 선택 
2. 라벨 생성 (최근 5년 존재하면 1(폐업) 아니면 0(운영)으로 라벨링 
3. 계층적 샘플링 (개방서비스명 + label 기준 train/test 분할) 
4. 결과 (80:20) 
- train 
 - 0 : 7372695
 - 1 : 1461417
- test 
 - 0 : 1843081
 - 1 : 365280 
- 클래스 불균형 처리-> 언더샘플링, 클래스 가중치 부여 등 필요

**Comment by khye00** [comment]
## XGBoost (샘플링으로 모델 1차)

### 1. 폐업 클래스 예측 (1)

| 샘플링 | Precision (정확히 맞춘 비율) | Recall (놓치지 않고 잡은 비율) | F1-score (종합) |
| --- | --- | --- | --- |
| 언더샘플링 | 0.3372 | **0.6946** | 0.4540 |
| 오버샘플링 | **0.3402** | 0.6898 | **0.4556** |
| 클래스가중치 | 0.3396 | 0.6908 | 0.4554 |
- 오버샘플링이 Precision, F1-score가 높은 편
- 언더샘플링은 Recall은 높지만 Precision이 낮음 → 다 맞추긴 했는데 오답도 많았다는 뜻
- 클래스 가중치는 중간 정도지만 안정적인 것으로 보임

### 2. 폐업아님(0) 클래스 예측 (운영 중)

| 샘플링 | Precision | Recall | F1-score |
| --- | --- | --- | --- |
| 언더샘플링 | **0.9162** | 0.7098 | 0.7999 |
| 오버샘플링 | 0.9156 | **0.7156** | **0.8033** |
| 클래스가중치 | 0.9157 | 0.7144 | 0.8027 |
- 오버샘플링과 클래스 가중치가 운영 중도 더 잘 맞추는 것으로 보임
- 언더샘플링은 운영 데이터 줄인 만큼 성능 살짝 감소

### 3. 전체 정확도 (Accuracy)

| 샘플링 | Accuracy |
| --- | --- |
| 언더샘플링 | 0.7071 |
| 오버샘플링 | **0.7110**  |
| 클래스가중치 | 0.7103 |
- 운영/폐업 예측 전체 정확도 기준에서는 오버샘플링이 가장 우수

### 평균 평가 지표 (macro avg / weighted avg)

| 샘플링 | Macro F1 | Weighted F1 |
| --- | --- | --- |
| 언더샘플링 | 0.6269 | 0.7392 |
| 오버샘플링 | **0.6295** | **0.7424** |
| 클래스가중치 | 0.6290 | 0.7418 |
- 오버샘플링이 폐업 클래스를 더 잘 예측하면서도 전체 예측 균형이 잘 맞음

**Comment by syleeie2310** [comment]
1) 피쳐 처리 방식 (운영방식 : 폐업인 경우는 다르게 처리 필요, 데이터 갱신 일자 등 검토 필요)
2) 카테고리 처리 방식 
---> 원핫 인코딩
3) 폐업 비율 유무 17.58%
4) 계층적 샘플링
--> 개방서비스명 / 폐업 유무 테이블

**Comment by khye00** [comment]
@syleeie2310 
멘토님 안녕하세요!
개방서비스명, 운영기간, 시도, 시군구를 원핫인코딩하여 XGBoost 모델을 돌리려고 하는데
pandas로 변환해서 진행하려고 하니 메모리 문제 때문에 오류가 발생합니다.

그래서 pyspark로 진행하려고 하니 VectorAssembler를 한 뒤에 모델을 돌려야 하는 것 같은데
서버리스에서는 안되는 것 같습니다...! 진행 방법 여쭙고 싶습니다..!

**Comment by syleeie2310** [comment]
네 내일 ML 클러스터 이야기 해보시죠. 피쳐 처리 방식 / 데이터 탐색부터 완료해주세요. Pandas에서 32기가 짜리 바꿧을때 가능한 수준으로 서버리스에서 해봐도 좋을거 같긴해요.

**Comment by khye00** [comment]
https://docs.google.com/spreadsheets/d/1PdwyPoQ0tG6cE957_4FER0Zub3ElPpr2nyAtpJK-bMM/edit?gid=0#gid=0

**Comment by khye00** [comment]
![Image](https://github.com/user-attachments/assets/83424681-ae33-4316-9143-3b175832a363)

**Comment by seoyoung1027** [comment]
### 사업장명 피처 
- 전처리 코드 
https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/653678358281676?o=639069795658224#command/653678358281679

- 임베팅 코드 
-https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/3364316723176646?o=639069795658224#command/8587842340612471

**Comment by khye00** [comment]
@syleeie2310 

멘토님 안녕하세요
70만개 데이터로 진행했습니다.

모델 학습한 데이터는 gold 카탈로그 ml_result_mois 스키마에 train, test 테이블 각각 생성했습니다!

- 작업한 노트북
[XGBOOST](https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/1657657003296827?o=639069795658224#command/8143215829744260)

**Comment by syleeie2310** [comment]
@khye00 혹시 모델 결과 추론 테이블은 train/test 합쳐서 전체 데이터에 적용 한번 해볼 수 있을까요?

할때마다 모델 학습하지 말고, 모델은 /Volumn 에다가 저장해놓고 불러오는 식으로 하는게 좋을 듯 해요.

아래 스키마에다가 볼륨 만들어서 모델은 거기다가 저장하심 될듯 해요.

gold.ml_result_mois

**Comment by khye00** [comment]
네 멘토님 해보겠습니다!

**Comment by khye00** [comment]
멘토님
/Volumes/gold/ml_result_mois/ml_result_file 경로에 학습한 모델 저장했습니다!

그리고 전체 데이터 적용은 현재 실행 중이며, 완료까지 시간이 조금 걸릴 것 같습니다!
ㄴ gold.ml_result_mois.model_full_data : 현재 진행 중인 테이블 입니다.

**Comment by syleeie2310** [comment]
폐업인 경우 27.2%

개방서비스명 / 폐업 유무 훈련 데이터 5백9십만개, 테스트 65만개 ---> 70만개 데이터 내에서 추가로 분류 (AUC : 76%)

시도/시군구/운영기간/업종구분명 ---> 개방서비스명, 시도, 시군구, 운영기간이 one-hot encoding, 업태구분명 라벨 encoding

model_full_data에 컬럼 추가해서 값만 업데이트 하면 됨

**Comment by syleeie2310** [comment]
1) gold.ml_result_mois.model_full_data 테이블을 만들기 
filtered_df 상태에서 컬럼을 예측값, 예측확률을 만들어두고 델타 테이블 저장

2) Merge 구문으로 데이터 청크로 잘라서 확률 저장하여 update 치는 방식으로 변경
-> final_df = spark.createDataFrame(pdf)
-> final_df.createOrReplaceTempView('final_df')

spark.sql("""
MERGE INTO  AS gold.ml_result_mois.model_full_data
USING final_df AS source <- chunk 해서 
ON target.unique = source.unique
WHEN MATCHED
THEN
  UPDATE SET target.예측값 = source.예측값,
             target.확률값 = source.확률값
""").show()

**Comment by khye00** [comment]
안녕하세요 멘토님

gold.ml_result_mois.model_full_data 테이블 작업 MERGE INTO 활용하여 완료했습니다.

70만개 데이터로 학습한 모델 결과 활용하여 진행했습니다.
(/Volumes/gold/ml_result_mois/ml_result_file/oversampling_model_70.pkl)

[작업한 노트북_행안부_전체 데이터 저장_gold.ml_result_mois.model_full_data](https://tacademykr-asacdataanalysis.cloud.databricks.com/editor/notebooks/598117186486212?o=639069795658224#command/8685880623855925)

**Comment by syleeie2310** [comment]
[참고 문서]
https://assets.docs.databricks.com/_extras/notebooks/source/mlflow/mlflow-end-to-end-example.html

https://docs.databricks.com/aws/en/mlflow/databricks-autologging
